File: 0159_Longest Substring with At Most Two Distinct Characters/LongestSubstringwithAtMostTwoDistinctCharacters.cs
public class Solution {
    public int LengthOfLongestSubstringTwoDistinct(string s) {
        var dict = new Dictionary<char, int>();
        var l = 0;
        var ans = 0;
 
        for (int r = 0; r < s.Length; r++)
        {
            if (!dict.ContainsKey(s[r]))
                dict.Add(s[r], 0);
            dict[s[r]]++;
 
            while (dict.Count > 2)
            {
                if (dict.ContainsKey(s[l]))
                {
                    dict[s[l]]--;
                    if (dict[s[l]] == 0)
                        dict.Remove(s[l]);
                    l++;
                }
            }
            ans = Math.Max(ans, r - l + 1);
        }
 
        return ans;
    }
}

File: 0159_Longest Substring with At Most Two Distinct Characters/LongestSubstringwithAtMostTwoDistinctCharacters.java
class Solution {
    public int lengthOfLongestSubstringTwoDistinct(String s) {
        int ans = 0;
        int l = 0;
        Map<Character, Integer> map = new HashMap();
        
        for(int r=0;r<s.length();r++){
            char c = s.charAt(r);
            if(map.containsKey(c)) {
                map.put(c, map.get(c)+1);
            }else{
                map.put(c, 1);
            }
            
            while(map.size() > 2) {
                char leftC = s.charAt(l++);
                if(map.containsKey(leftC)) {
                    map.put(leftC, map.get(leftC) - 1);
                    if(map.get(leftC) == 0)
                        map.remove(leftC);
                }
            }
            
            if(map.size() <= 2) {
                ans = Math.max(ans, r - l + 1);
            }
        }
        
        return ans;
    }
}